routing/routing.go

```diff
@@ -140,10 +142,83 @@ type RouteFilter struct {
 	Index int
 }
 
+// LBMetrics contains metrics used by LB algorithms
+type LBMetrics struct {
+	inflightRequests int
+	mutex            sync.RWMutex
+}
+
+// IncInflightRequest increments the number of outstanding requests from the proxy to a given backend.
+func (m *LBMetrics) IncInflightRequest() {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	m.inflightRequests++
+}
+
+// DecInflightRequest decrements the number of outstanding requests from the proxy to a given backend.
+func (m *LBMetrics) DecInflightRequest() {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	m.inflightRequests--
+}
+
+// GetInflightRequests decrements the number of outstanding requests from the proxy to a given backend.
+func (m *LBMetrics) GetInflightRequests() int {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+	return m.inflightRequests
+}
+
 // LBEndpoint represents the scheme and the host of load balanced
 // backends.
 type LBEndpoint struct {
 	Scheme, Host string
+	Metrics      *LBMetrics
+}
+
+type LBEndpointCollection struct {
+	endpoints    []LBEndpoint
+	endpointsMap map[string]LBEndpoint
+}
+
+// LBEndpointCollection represents a collection of endpoints.
+// The underlying endpoints can be accessed using public getters and setters.
+// Internally endpoints are stored in two synchronized data structures, so
+// it is possible to reference them by index or key.
+func NewEndpointCollection(r *Route) (*LBEndpointCollection, error) {
+	collection := &LBEndpointCollection{
+		endpoints:    make([]LBEndpoint, 0, len(r.Route.LBEndpoints)),
+		endpointsMap: map[string]LBEndpoint{},
+	}
+	for _, e := range r.Route.LBEndpoints {
+		eu, err := url.ParseRequestURI(e)
+		if err != nil {
+			return nil, err
+		}
+		collection.Set(eu.Host, LBEndpoint{Scheme: eu.Scheme, Host: eu.Host, Metrics: &LBMetrics{}})
+	}
+	return collection, nil
+}
+
+// Stores an endpoint indexed by key
+func (e *LBEndpointCollection) Set(key string, item LBEndpoint) {
```

Comment:
As `LBEndpointCollection` seems to be immutable datastructure, should this be private or even inlined?

Reply:
Totally, I just replaced the existing data structure but I will change this tomorrow.